pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+import sys
+from typing import (
+    Any,
+    Generic,
+    Protocol,
+    TypeVar,
+    Union,
+    overload,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    Timedelta,
+    Timestamp,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+VALID_CLOSED: frozenset
+
+OrderableScalarT = TypeVar("OrderableScalarT", int, float)
+OrderableTimesT = TypeVar("OrderableTimesT", Timestamp, Timedelta)
+OrderableT = TypeVar("OrderableT", int, float, Timestamp, Timedelta)
+
+class IntervalMixinProtocol(Protocol): ...
+
+class _LengthDescriptor:
+    @overload
+    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
+    @overload
+    def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> Timedelta: ...
+
+class IntervalMixin:
+    @property
+    def closed_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def closed_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def mid(self: IntervalMixinProtocol) -> float: ...
```

Comment:
`mid` should be generic. `length` should also be here, it is implemented in `IntervalMixin`, not in `Interval` (this will craete some [headaches](https://github.com/pandas-dev/pandas/pull/44922/files#diff-0a83f799817ae13154c34067c21d5a126efb6b433ec0860d88140a85751f8735R93) - because `IntervalTree` also inherits from  `IntervalMixin`).

I will not work on my PR for the next days, hopefully yours makes it. Later this week I have more time to pick up my PR if needed.

Reply:
fixed in next commit